PopPUNK/network.py

```diff
@@ -1226,49 +1316,84 @@ def generate_minimum_spanning_tree(G, from_cugraph = False):
         else:
             sys.stderr.write("generate_minimum_spanning_tree requires a weighted graph\n")
             raise RuntimeError("MST passed unweighted graph")
-    else:
-        mst_network = G
 
     # Find seed nodes as those with greatest outdegree in each component
+    num_components = 1
     seed_vertices = set()
-    component_assignments, component_frequencies = gt.label_components(mst_network)
-    for component_index in range(len(component_frequencies)):
-        component_members = component_assignments.a == component_index
-        component = gt.GraphView(mst_network, vfilt = component_members)
-        component_vertices = component.get_vertices()
-        out_degrees = component.get_out_degrees(component_vertices)
-        seed_vertex = list(component_vertices[np.where(out_degrees == np.amax(out_degrees))])
-        seed_vertices.add(seed_vertex[0]) # Can only add one otherwise not MST
-
-    # If multiple components, calculate distances between seed nodes
-    if len(component_frequencies) > 1:
-        # Extract distances
-        connections = []
-        max_weight = float(np.max(G.edge_properties["weight"].a))
-        for ref in seed_vertices:
-            seed_edges = G.get_all_edges(ref, [G.ep['weight']])
-            found = False  # Not all edges may be in graph
-            for seed_edge in seed_edges:
-                if seed_edge[1] in seed_vertices:
-                    found = True
-                    connections.append((seed_edge))
-            # TODO: alternative would be to requery the DB (likely quick)
-            if found == False:
-                for query in seed_vertices:
-                    if query != ref:
-                        connections.append((ref, query, max_weight))
+    if from_cugraph:
+        mst_df = cugraph.components.connectivity.connected_components(mst_network)
+        num_components_idx = mst_df['labels'].max()
+        num_components = mst_df.iloc[num_components_idx]['labels']
+        if num_components > 1:
+            mst_df['degree'] = mst_network.in_degree()['degree']
+            # idxmax only returns first occurrence of maximum so should maintain
+            # MST - check cuDF implementation is the same
+            max_indices = mst_df.groupby(['labels'])['degree'].idxmax()
+            seed_vertices = mst_df.iloc[max_indices]['vertex']
+            num_components = seed_vertices.size()
+        del mst_df
```

Comment:
Shouldn't this happen automatically at the end of the function? Or does it potentially use up loads of memory between here and the end of the function?

Reply:
Removed in 8fc9763.